### PR TITLE
Add CA Search functionality for the windows cert store

### PR DESCRIPTION
**Pull request by ColinSullivan1**
This PR adds a new field for searching the windows Certificate Store for CA certificates.

When specifying a value in the `ca_certs_match` field, the following locations are searched:
- Trusted Root Certification Authorities
- Third-Party Root Certification Authorities
- Intermediate Certification Authorities

Based on user input, the first matching certificate from subject in the list, from each location, is added to the certificate pool.  If no certificates can be found an error is reported.

Example usage:

```text
tls {
  cert_store: "WindowsLocalMachine"
  cert_match_by: "Subject"
  cert_match: "client001"
  ca_certs_match: ["CA1","CA2"]
  timeout: 3s
}
```

Additional Notes:
Test scripts have been added to add and remove the test CA certificate, and existing windows certificates have been updated as they were expired.

Signed-off-by: Colin Sullivan <colin@luxantsolutions.com>

### CODE REVIEW

File: server/certstore/certstore_windows.go
Comment (bruth) [inline server/certstore/certstore_windows.go:156]: No need to allocate a slice unless a known fixed size is desired, the first `append` will do the allocation.

File: server/certstore/certstore_windows.go
Comment (bruth) [inline server/certstore/certstore_windows.go:166]: No need for a `count`, doing a `len(errs)` would suffice.

File: server/certstore/certstore_windows.go
Comment (bruth) [inline server/certstore/certstore_windows.go:222]: Would be safer to do `if len(caCertsMatch) != 0` which works on `nil` slices.

File: server/certstore/certstore_windows.go
Comment (bruth) [inline server/certstore/certstore_windows.go:362]:
```suggestion
// caCertBySubject matches and returns all matching certificates of the subject field.
```

File: server/certstore/certstore_windows.go
Comment (bruth) [inline server/certstore/certstore_windows.go:118]: For posterity, are there any further docs that can be linked to in a comment in case we need to dig into understanding the semantics of these various stores in the future?

File: server/certstore/certstore_windows.go
Comment (bruth) [inline server/certstore/certstore_windows.go:380]: Can be initialized under the `var` block above as `rv = []*x509.Certificate`

File: server/certstore/certstore_windows.go
Comment (bruth) [inline server/certstore/certstore_windows.go:386]: Since only this error value is being checked, what are the other error values that are ignored here?

File: server/certstore_windows_test.go
Comment (bruth) [inline server/certstore_windows_test.go:67]: Should this be deleted if commented out?

**Review by bruth — CHANGES_REQUESTED** [review]
Surface level pass on functionality (which looks good!) and a few basic comments.

Not for this PR, but I think we need to carve out Windows-specific tests in a GitHub Action job.

File: server/certstore/errors.go
Comment (ripienaar) [inline server/certstore/errors.go:72]: seems unused

File: test/configs/certs/tlsauth/certstore/server.p12
Comment (wallyqs) [inline test/configs/certs/tlsauth/certstore/server.p12:1]: (just checking, do we need to keep this file?)

File: test/configs/certs/tlsauth/certstore/update_keys.bat
Comment (wallyqs) [inline test/configs/certs/tlsauth/certstore/update_keys.bat:1]: is this file needed?

File: server/certstore/certstore_windows.go
Comment (ColinSullivan1) [reply server/certstore/certstore_windows.go:222]: Will fix, thanks Byron.

File: server/certstore/certstore_windows.go
Comment (ColinSullivan1) [reply server/certstore/certstore_windows.go:166]: Agree - that's cleaner - if len(errs) == len(caCertsMatch)...

File: server/certstore/certstore_windows.go
Comment (ColinSullivan1) [reply server/certstore/certstore_windows.go:118]: I'll add links to the MSFT docs.

File: server/certstore_windows_test.go
Comment (ColinSullivan1) [reply server/certstore_windows_test.go:67]: It's actually equivalent now with the CA search field above, so I moved it up a few lines.  I can add a comment.

File: server/certstore/certstore_windows.go
Comment (ColinSullivan1) [reply server/certstore/certstore_windows.go:174]: Will do, thanks R.I.  Not sure how to get around the breaking change unless I create a struct to pass.  I'll update the PR comments to this effect.

Didn't realize that'd break the CLI. Would you like me to make corresponding changes there after this PR?

File: server/certstore/errors.go
Comment (ColinSullivan1) [reply server/certstore/errors.go:72]: I'll fix that - thanks!

File: server/certstore/certstore_windows.go
Comment (ColinSullivan1) [reply server/certstore/certstore_windows.go:156]: Sounds good - will fix.

**Comment by bruth** [comment]
My suggestion for this one is for @derekcollison to do a review. Once its approved and merged, we can include in a preview.2 release of 2.11 for folks to test.

**Comment by ripienaar** [comment]
> I doubt I am the right person to do this review.

feature wise this seems fine to me - pity about the breaking API change but for this partcular code I dont think its a problem.

**Comment by bruth** [comment]
Agree with @ripienaar. @derekcollison Then let's approve and merge once green, cut a new preview shortly, and then iterate based on feedback.

**Comment by bruth** [comment]
The failure appears to be some change in CLI arguments for MQTT testing. I don't see it necessary to rebase for that test to pass, so approve and merge when ready.

**Comment by ripienaar** [comment]
@ColinSullivan1 I am not seeing here a test that shows a client loading CA from certstore can connect to a server thats same - we do have these https://github.com/nats-io/nats-server/blob/6e768e8481ea13482428c70f90c21d842721d671/server/certstore_windows_test.go#L227 but none with CA loaded from a certstore.

I am having some trouble getting this scenario going, so a test showing its possible would be great!

**Comment by ripienaar** [comment]
To make it work I think we need to add here https://github.com/nats-io/nats-server/blob/6e768e8481ea13482428c70f90c21d842721d671/server/certstore/certstore_windows.go#L235

```
			config.RootCAs = caPool
```

But I am not sure if thats the right thing to do? I can also do it on my side basically swap those 2 for client purposes?

**Comment by ripienaar** [comment]
So for me to use it in a client without modifying the server, it would be:

```
err = certstore.TLSConfig(storeType, matchBy, c.config.WinCertStoreMatch, c.config.WinCertStoreCaMatch, tlsc)
	if err != nil {
		return nil, err
	}

	if tlsc.ClientCAs != nil {
		tlsc.RootCAs = tlsc.ClientCAs
		tlsc.ClientCAs = nil
	}
```

I am not sure whats the right thing to do though

**Comment by ripienaar** [comment]
Seems we do this already in the server here:

https://github.com/nats-io/nats-server/blob/f1cd3ed141b0f5fe0fe62004c0988a8c829bdc33/server/opts.go#L1398-L1400

So I am ok doing that unless you have other oppionions @ColinSullivan1

**Comment by ColinSullivan1** [comment]
> but none with CA loaded from a certstore.

Works for me...

**Comment by ColinSullivan1** [comment]
> @ColinSullivan1 I am not seeing here a test that shows a client loading CA from certstore can connect to a server thats same - we do have these
> 
> https://github.com/nats-io/nats-server/blob/6e768e8481ea13482428c70f90c21d842721d671/server/certstore_windows_test.go#L227
> 
> but none with CA loaded from a certstore.
> I am having some trouble getting this scenario going, so a test showing its possible would be great!

Thanks R.I. - I'll take a look. The CAs added to the certstore were generated from the same CA the client should be using, but I understand we want to future-proof the tests. I'll aim for today, or if not later next week.

**Comment by ColinSullivan1** [comment]
> > @ColinSullivan1 I am not seeing here a test that shows a client loading CA from certstore can connect to a server thats same - we do have these
> > https://github.com/nats-io/nats-server/blob/6e768e8481ea13482428c70f90c21d842721d671/server/certstore_windows_test.go#L227
> > 
> > but none with CA loaded from a certstore.
> > I am having some trouble getting this scenario going, so a test showing its possible would be great!
> 
> Thanks R.I. - I'll take a look. The CAs added to the certstore were generated from the same CA the client should be using, but I understand we want to future-proof the tests. I'll aim for today, or if not later next week.

Per our slack convo I'll hold off on this. lmk if there are any other updates/changes - happy to make them.

**Pull request merged**
